# Tarea 1 Bases de Datos 2023-1 Grupo 23

## Integrantes:
* Carolina Sire -Rol-202173105-9
* Eva Wang - Rol-202111004-6

## Requisitos para Correr el Notebook

Para poder utilizar este Notebook se debe tener instalado:
- **Python 3**
- **Jupyter**
- La librería **ipython-sql** (Poder ejecutar sql en Jupyter Notebook)
- PostgreSQL y su conector desde Python(**psycopg2**)

**IMPORTANTE:** Este Jupyter Notebook se conectará a su servicio de Base de Datos PostgreSQL de manera **local**, por lo que **no funcionará correctamente en el entorno Google Collab.**

Para instalar las dependencias, ejecute (**recuerde actualizar pip3**):

In [ ]:
! pip3 install ipython-sql
! pip3 install psycopg2

#Use to generate names
! pip3 install names

## Importar dependencias

In [2]:
from datetime import datetime
import random as r
import names

#Librerías para PostgreSQL
import psycopg2

Para conectarse a la base de datos, completen los datos relevantes en la variable DATABASE_URL.

* Cambie `user` por el usuario que utilizó en Postgres para acceder a la Base de Datos. Por defecto `postgres`.
* Cambie `password` por la contraseña de dicho usuario. **Recuerde borrarla antes de entregar su tarea.**
* Cambie `port` por el puerto en que se ejecuta su servicio de Base de Datos. Por defecto `5432`.
* Cambie `database` por el nombre de la BD (Schema) que creó en Postgres.

Ejemplo: `postgresql://postgres:pass_segura@localhost:5432/tarea-bd`

In [3]:
%reload_ext sql
%env DATABASE_URL=postgresql://postgres:1972@localhost:5432/tarea1

env: DATABASE_URL=postgresql://postgres:1972@localhost:5432/tarea1


Para conectarse de forma declarativa a la Base de Datos, utilice el siguiente código:

In [4]:
#Recuerde cambiar los valores de USER, DBNAME y PASSWORD por los mismos de la celda superior.
connection = psycopg2.connect(
   host="localhost",
   user="postgres",
   dbname='tarea1',
   password="1972"
)

print(connection)
cursor = connection.cursor()

<connection object at 0x00000150728EEF00; dsn: 'user=postgres password=xxx dbname=tarea1 host=localhost', closed: 0>


## Crear tablas (36 pts.)

Aquí inserta el código SQL que creara las tablas de la base de datos:

### Tabla "facciones" (6 pts)

In [29]:
%%sql
CREATE TABLE facciones (
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    nombre VARCHAR (45) NOT NULL,
    descripcion VARCHAR (45) NOT NULL,
    PRIMARY KEY (id)
);

 * postgresql://postgres:***@localhost:5432/tarea1
Done.


[]

### Tabla "trabajos" (6 pts)

In [30]:
%%sql
CREATE TABLE trabajos (
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    descripcion VARCHAR(45) NOT NULL,
    sueldo INTEGER NOT NULL,
    PRIMARY KEY (id)
);

 * postgresql://postgres:***@localhost:5432/tarea1
Done.


[]

### Tabla "personas" (6 pts)

In [31]:
%%sql
CREATE TABLE personas (
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    apellidos VARCHAR(45) NOT NULL,
    fecha_nacimiento DATE NOT NULL,
    infectado BOOL NOT NULL,
    conyugue INTEGER,
    PRIMARY KEY (id),
    FOREIGN KEY (conyugue) REFERENCES personas (id)
);

 * postgresql://postgres:***@localhost:5432/tarea1
Done.


[]

### Tabla "cuentas_bancarias" (6 pts)

In [32]:
%%sql
CREATE TABLE cuentas_bancarias (
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    tipo_de_cuenta VARCHAR(45) NOT NULL,
    banco VARCHAR(45) NOT NULL,
    saldo INTEGER NOT NULL,
    id_persona INTEGER,
    PRIMARY KEY (id),
    FOREIGN KEY (id_persona) REFERENCES personas (id)
);

 * postgresql://postgres:***@localhost:5432/tarea1
Done.


[]

### Tabla "persona_pertenece_faccion" (6 pts)

In [33]:
%%sql
CREATE TABLE persona_pertenece_faccion (
    id_persona INTEGER,
    id_faccion INTEGER,
    PRIMARY KEY (id_persona, id_faccion),
    FOREIGN KEY (id_persona) REFERENCES personas (id),
    FOREIGN KEY (id_faccion) REFERENCES facciones (id)
);

 * postgresql://postgres:***@localhost:5432/tarea1
Done.


[]

### Tabla "persona_tiene_trabajo" (6 pts)

In [34]:
%%sql
CREATE TABLE persona_tiene_trabajo (
    id_persona INTEGER,
    id_trabajo INTEGER,
    estado BOOL NOT NULL,
    ultima_vez_realizado TIMESTAMP NOT NULL,
    PRIMARY KEY (id_persona, id_trabajo),
    FOREIGN KEY (id_persona) REFERENCES personas (id),
    FOREIGN KEY (id_trabajo) REFERENCES trabajos (id)
);


 * postgresql://postgres:***@localhost:5432/tarea1
Done.


[]

## Código Generador de Datos

Ejecute este código para cargar datos a las tablas creadas.

PD: Si el codigo tira error por "violar la unicidad de las PK" es por una coincidencia de numeros random, solamente volver a ejecutar el codigo

In [35]:
#Recuerde haber ejecutado todas las celdas anteriores!

# Generador de timestamp aleatorio.
def generate_timestamp_nacimiento():
    year = r.randint(1970,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

def generate_timestamp_trabajo():
    year = r.randint(2019,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

print("Borrando Datos Antiguos...")

# Limpia la base de datos.
connection.commit() # En caso de algun error
query = 'TRUNCATE TABLE "facciones", "trabajos", "personas", "cuentas_bancarias", "persona_pertenece_faccion", "persona_tiene_trabajo" RESTART IDENTITY'
cursor.execute(query)
connection.commit()

print("Generando Datos...")

facciones = [("FEDRA", "La Agencia Federal de Respuesta a Desastres"), ("Luciérnagas", "Grupo de milicias revolucionario"), ("Carroñeros", "Sin identificación"), ("Pythonia", "Grupo defensor de la libertad"), ("MaCdonia", "Los exiliados de la carceles"), ("JavaScriptia", "Grupo defensor de la paz")]
trabajos = [("Limpieza", "Encargado de limpiar las calles"),
("Limpiador de cloacas","Encargado de limpiar las cloacas"),
("Militar","Encargado de proteger la ciudad"),
("Explorador","Encargado de explorar la ciudad"),
("Médico","encargado de curar a los enfermos"),
("Profesor","Encargado de enseñar a los niños"),
("Cremador","Encargado de quemar los cadáveres"),
("Científico","Encargado de investigar"),
("Cocinero","Encargado de cocinar"),
("Carpintero","Encargado de construir"),
("Cerrajero","Encargado de reparar las cerraduras"),
("Cartero","Encargado de repartir los correos"),
("Cajero","Encargado de cobrar los impuestos")]
tipos_de_cuenta_bancaria = ["Cuenta Vista", "Cuenta Ahorro", "Cuenta Corriente", "Cuenta RUT", "Cuenta de Inversion", "Cuenta PrePago", "Cuenta de Credito"]
bancos = ["FedraBank S.A", "Banco Pedrito Pascal SpA", "Banco Pythonia S.A", "Banco de la Nación S.A", "Banco Santander SpA", "Banco BCI SpA", "Banco Estado SpA", "Banco Consorcio S.A", "Banco BBVA", "Banco Scotiabank"]

# Variables.
n_facciones = len(facciones)
n_trabajos = len(trabajos)
n_personas = 1000
n_cuentas_bancarias = 120
n_persona_pertenece_faccion = 300
n_persona_tiene_trabajo = 300
n_casados = 100

# Genera personas.
personas = []
for _ in range(n_personas):
    nombre = names.get_first_name()
    apellidos = names.get_last_name()+" "+names.get_last_name()
    fecha_nacimiento = generate_timestamp_nacimiento()
    infectado = r.choice([True, False])
    
    personas.append((nombre, apellidos, fecha_nacimiento, infectado, None))


# Genera cuentas bancarias.
cuentas_bancarias = []
for _ in range(n_cuentas_bancarias):
    tipo_de_cuenta = r.choice(tipos_de_cuenta_bancaria)
    banco = r.choice(bancos)
    saldo = r.randint(0, 10000)
    id_persona = r.choice(range(1, n_personas+1))
    
    cuentas_bancarias.append((tipo_de_cuenta, banco, saldo, id_persona))
    
    
# Generate persona_pertenece_faccion.
persona_pertenece_faccion = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_facciones+1))) for _ in range(1, n_persona_pertenece_faccion+1)])

# Generate persona_tiene_trabajo.
persona_tiene_trabajo_prev = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_trabajos+1))) for _ in range(1, n_persona_tiene_trabajo+1)])
persona_tiene_trabajo = [(id_persona, id_trabajo, r.choice([True, False]), generate_timestamp_trabajo()) for id_persona, id_trabajo in persona_tiene_trabajo_prev]
    
# SQL Statements.
inserts = [
    'INSERT INTO "facciones" ("nombre", "descripcion") VALUES (%s, %s);',
    'INSERT INTO "trabajos" ("nombre", "descripcion", "sueldo") VALUES (%s, %s, %s);',
    'INSERT INTO "personas" ("nombre", "apellidos", "fecha_nacimiento", "infectado", "conyugue") VALUES (%s, %s, %s, %s, %s);',
    'INSERT INTO "cuentas_bancarias" ("tipo_de_cuenta", "banco", "saldo", "id_persona") VALUES (%s, %s, %s, %s);',
    'INSERT INTO "persona_pertenece_faccion" ("id_persona", "id_faccion") VALUES (%s, %s);',
    'INSERT INTO "persona_tiene_trabajo" ("id_persona", "id_trabajo", "estado", "ultima_vez_realizado") VALUES (%s, %s, %s, %s);'
]

for faccion in facciones:
    cursor.execute(inserts[0], faccion)

for trabajo in trabajos:
    t = (trabajo[0], trabajo[1], r.randint(100, 5000))
    cursor.execute(inserts[1], t)

# Crear una lista con personas casadas y otra con personas solteras. donde se cambie la ide de la persona que se casó.

for persona in personas:
    cursor.execute(inserts[2], persona)
    
for cb in cuentas_bancarias:
    cursor.execute(inserts[3], cb)

for ppf in persona_pertenece_faccion:
    cursor.execute(inserts[4], ppf)
    
for ptt in persona_tiene_trabajo:
    cursor.execute(inserts[5], ptt)
    
connection.commit()


# Generar Matrimonios
n_repetidos = []

for i in range(n_casados):
    random = r.randint(1, n_personas)
    if random in n_repetidos:
        continue
    else:
        random2 = r.randint(1, n_personas)
        if random2 in n_repetidos:
            continue
        else:
            n_repetidos.append(random)
            n_repetidos.append(random2)
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random2, random))
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random, random2))
            connection.commit()

print("Todo Listo!")

Borrando Datos Antiguos...
Generando Datos...
Todo Listo!


## Consultas (64 pts.)

Escriba aquí las consultas pedidas. Recuerde agregar una descripción _breve_ de cómo funciona la solución. **Escriba las consultas de forma ordenada.**

**NOTA:** Cuando en una consulta diga "X cosa" significa que ustedes deben de elegir ese X por lo que para cualquier otro X debe de funcionar de igual forma.

### Consulta 1 (6 pts)
**ID, nombre y primer apellido de los habitantes de FEDRA**

Ejemplo:

|ID |  Nombre      |  Apellido  |
|---|-----------|-----------|
| 12 |Joel | Miller |
|21 | Ellie | Williams |
| ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._


In [14]:
%%sql
SELECT
    personas.id "ID", personas.nombre "Nombre" , personas.apellidos "Apellido"
FROM personas
    INNER JOIN persona_pertenece_faccion ON personas.id = persona_pertenece_faccion.id_persona
    INNER JOIN facciones ON facciones.id = persona_pertenece_faccion.id_faccion
WHERE facciones.nombre = 'FEDRA';

 * postgresql://postgres:***@localhost:5432/tarea1
44 rows affected.


[(58, 'Marcus', 'Flippen Lopez'),
 (137, 'George', 'Greene Jordan'),
 (971, 'Leroy', 'Vanhoose Holm'),
 (575, 'Jayson', 'Kotterna Tran'),
 (273, 'Frances', 'Russell Jackson'),
 (973, 'James', 'Swanson Stayton'),
 (205, 'Helen', 'Ford Mccune'),
 (990, 'Lucille', 'Cortez Williams'),
 (739, 'Donald', 'Calicott Stiltner'),
 (895, 'Cecile', 'Donald Salas'),
 (314, 'James', 'Cavin Barth'),
 (305, 'Alice', 'Thomas Harrison'),
 (67, 'Todd', 'Parker Anderson'),
 (78, 'Nicholas', 'Cosby Moore'),
 (967, 'Nicole', 'Cauthen Johnson'),
 (729, 'Erica', 'Farr More'),
 (333, 'Doris', 'Bland Abeles'),
 (71, 'Paul', 'Geiser Pelletier'),
 (284, 'Jerry', 'Harris Costales'),
 (605, 'Jill', 'Erb Gibbs'),
 (72, 'Matthew', 'Shelby Jurczyk'),
 (140, 'Curtis', 'Shiba Pfaff'),
 (512, 'Robert', 'Flannigan Palos'),
 (342, 'Dennis', 'Price Braatz'),
 (835, 'Dinah', 'Carroll Medina'),
 (493, 'Floyd', 'Oneal Casey'),
 (966, 'Nicholas', 'Proulx Schwartz'),
 (332, 'Andrew', 'Scott Looney'),
 (46, 'Jessica', 'Monford Pryor'),
 (376, 'Nancy', 'Hawkins Corcoran'),
 (215, 'Thomas', 'Plessinger Delgado'),
 (338, 'Elsie', 'Ramirez Condon'),
 (483, 'Charles', 'Curry Taylor'),
 (287, 'Jody', 'Ulmer Hashim'),
 (179, 'Stephanie', 'Dukes Coleman'),
 (606, 'Craig', 'Hope York'),
 (940, 'Louis', 'Miller Dunn'),
 (889, 'Daniel', 'Sexton Betts'),
 (356, 'Ernest', 'Huggins Schmidt'),
 (15, 'David', 'Huelle Perez'),
 (52, 'Julie', 'Mariner Camp'),
 (547, 'Ralph', 'Cook Owens'),
 (637, 'Alejandro', 'Matin Cahill'),
 (241, 'Thomas', 'Kelty Fletcher')]

### Consulta 2 (6 pts)
**Cantidad de personas nacidas después del inicio de la pandemia.**(inicio de la pandemia 01-12-2019)

Ejemplo:


| Cantidad     | 
|--------------|
| 292 | 

**Solución:** _Escriba aquí la descripción de su solución._

In [12]:
%%sql
SELECT
    COUNT(id) "Cantidad"
FROM
    personas
WHERE
    personas.fecha_nacimiento BETWEEN '2019-12-01' AND CURRENT_DATE;

 * postgresql://postgres:***@localhost:5432/tarea1
1 rows affected.


[(52,)]

### Consulta 3 (6 pts)
**ID, nombre, apellidos y saldo total de todas las personas**( Limitar a 20 filas para visualizar)

Ejemplo:

| ID    | Nombre | Apellido | Dinero | 
|----------------|--------|-|-|
| 13 | Tommy | Miller | 12501 |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [16]:
%%sql
SELECT
    personas.id "ID", personas.nombre "Nombre", personas.apellidos "Apellido", cuentas_bancarias.saldo "Dinero"
FROM 
    personas INNER JOIN cuentas_bancarias
ON
    personas.id = cuentas_bancarias.id_persona
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/tarea1
20 rows affected.


[(44, 'Benjamin', 'Block Long', 2716),
 (46, 'Jessica', 'Monford Pryor', 3601),
 (63, 'Robert', 'Bergeron Leach', 1003),
 (74, 'Benjamin', 'Collazo Motl', 3884),
 (77, 'Dana', 'Herrera Locklear', 8009),
 (86, 'Armando', 'Doherty Kimball', 8508),
 (89, 'Michael', 'Bronstein West', 3539),
 (92, 'Paul', 'Mccarty Laguna', 5675),
 (94, 'Ashley', 'Millette Davenport', 580),
 (99, 'Helen', 'Lichty Sheridan', 4799),
 (106, 'Julie', 'Jiles Killen', 7800),
 (108, 'Bryan', 'Williams Lattea', 6726),
 (117, 'Theresa', 'Dewitt Owens', 6581),
 (124, 'Gloria', 'Clayton Parson', 4586),
 (121, 'Ellen', 'Marcinkiewicz Estelle', 4266),
 (65, 'Victoria', 'Williams Garnick', 5514),
 (133, 'Brittany', 'Laitila Walker', 6303),
 (158, 'Kenneth', 'Topps Marsh', 6598),
 (170, 'Veronica', 'Johnson Rhoden', 807),
 (173, 'Jimmy', 'Aldridge Beatty', 3526)]

### Consulta 4 (6 pts)
**Nombre de la Facción y la cantidad de trabajadores activos hasta fecha X.**

Ejemplo:

| Facción     | Cantidad de trabajadores |
|----------------|--------|
| FEDRA | 1273 |
| Luciérnagas | 121 |
| Carroñeros  | 72 |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [13]:
%%sql
SELECT
    facciones.nombre "Facción", COUNT(id_persona) "Cantidad de trabajadores"
FROM 
    facciones, persona_pertenece_faccion
GROUP BY
    facciones.nombre;

 * postgresql://postgres:***@localhost:5432/tarea1
6 rows affected.


[('JavaScriptia', 293),
 ('Luciérnagas', 293),
 ('Pythonia', 293),
 ('MaCdonia', 293),
 ('FEDRA', 293),
 ('Carroñeros', 293)]

### Consulta 5 (6 pts)
**Nombre de la Facción y su cantidad de personas infectadas.**

Ejemplo:

|  Facción  | Personas infectadas | 
|--------|-------|
| FEDRA | 251 |
| Luciérnagas | 17 |
| Carroñeros  | 12 |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [21]:
%%sql
SELECT
    facciones.nombre "Facción", COUNT(personas.infectado) "Personas infectadas"
FROM 
    persona_pertenece_faccion
    INNER JOIN facciones ON persona_pertenece_faccion.id_faccion = facciones.id
    INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id
WHERE
    infectado = TRUE
GROUP BY
    facciones.nombre;

 * postgresql://postgres:***@localhost:5432/tarea1
6 rows affected.


[('JavaScriptia', 30),
 ('Luciérnagas', 26),
 ('Pythonia', 27),
 ('MaCdonia', 22),
 ('FEDRA', 13),
 ('Carroñeros', 20)]

### Consulta 6 (6 pts)
**Nombre de la Facción y su trabajo más realizado.**

Ejemplo:

|  Facción  | Trabajo | 
|--------|-------|
| FEDRA | Limpiador de cloacas |
| Luciérnagas | Médico |
| Carroñeros  | Explorador |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [23]:
%%sql
SELECT
    facciones.nombre "Facción", MAX(trabajos.nombre) "Trabajo" 
FROM
    personas
    INNER JOIN persona_tiene_trabajo ON personas.id= persona_tiene_trabajo.id_persona
    INNER JOIN trabajos ON trabajos.id= persona_tiene_trabajo.id_trabajo
    INNER JOIN persona_pertenece_faccion ON persona_pertenece_faccion.id_persona = personas.id
    INNER JOIN facciones ON facciones.id = persona_pertenece_faccion.id_faccion
GROUP BY facciones.nombre;

 * postgresql://postgres:***@localhost:5432/tarea1
6 rows affected.


[('JavaScriptia', 'Profesor'),
 ('Luciérnagas', 'Profesor'),
 ('Carroñeros', 'Profesor'),
 ('Pythonia', 'Profesor'),
 ('MaCdonia', 'Militar'),
 ('FEDRA', 'Profesor')]

### Consulta 7 (8 pts)
**Nombre de la Facción y su Saldo total** (Es decir, la suma total de los saldos de cada persona perteneciente a la facción.)

Ejemplo:

|Facción | Saldo |
|--------|-|
| FEDRA | 523000 |
| Luciérnagas | 130023 |
| Carroñeros  | 17432 |
| ...  | ...|

**Solución:** _Escriba aquí la descripción de su solución._

In [24]:
%%sql
SELECT
    facciones.nombre "Facción", SUM(cuentas_bancarias.saldo) "Saldo"
FROM 
    personas
    INNER JOIN persona_pertenece_faccion ON persona_pertenece_faccion.id_persona = personas.id
    INNER JOIN facciones ON facciones.id = persona_pertenece_faccion.id_faccion
    INNER JOIN cuentas_bancarias ON personas.id = cuentas_bancarias.id_persona
GROUP BY
    facciones.nombre;

 * postgresql://postgres:***@localhost:5432/tarea1
6 rows affected.


[('JavaScriptia', 36508),
 ('Pythonia', 48770),
 ('MaCdonia', 24587),
 ('FEDRA', 25967),
 ('Luciérnagas', 58039),
 ('Carroñeros', 19507)]

### Consulta 8 (6 pts)
**Nombre del Banco, cantidad de usuarios pertenecientes al banco y saldo total del banco.**

###### Es decir, todos los bancos que ocupan los habitantes de cada faccion

Ejemplo:

|Banco | Cantidad|Saldo|
|----|---|---|
|FedraBank S.A |321|20000|
|Banco Pedrito Pascal SpA  |21|200| 
| Banco De Chile|213|56000|
| ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [26]:
%%sql
SELECT
    cuentas_bancarias.banco "Banco", COUNT(cuentas_bancarias.id) "Cantidad", SUM(cuentas_bancarias.saldo) "Saldo"
FROM 
    cuentas_bancarias
GROUP BY
    cuentas_bancarias.banco;

 * postgresql://postgres:***@localhost:5432/tarea1
10 rows affected.


[('Banco Consorcio S.A', 10, 55565),
 ('Banco de la Nación S.A', 17, 86234),
 ('Banco Santander SpA', 13, 73737),
 ('Banco Pythonia S.A', 5, 13164),
 ('Banco Estado SpA', 10, 41671),
 ('Banco BCI SpA', 14, 77895),
 ('FedraBank S.A', 13, 42923),
 ('Banco BBVA', 9, 46168),
 ('Banco Pedrito Pascal SpA', 11, 62982),
 ('Banco Scotiabank', 18, 72176)]

### Consulta 9 (8 pts)
**Nombre de la Facción y el Banco más utilizado por sus habitantes**(Si hay empate, que se muestre solo uno.)

Ejemplo:
   
|Facción | Banco |
| - | - |
| FEDRA | FedraBank S.A |
| Luciérnagas | Banco Pedrito Pascal SpA |
| ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [28]:
%%sql
SELECT
    facciones.nombre "Facción", MAX(cuentas_bancarias.banco) "Banco" 
FROM
    personas
    INNER JOIN persona_pertenece_faccion ON persona_pertenece_faccion.id_persona = personas.id
    INNER JOIN facciones ON facciones.id = persona_pertenece_faccion.id_faccion
    INNER JOIN cuentas_bancarias ON cuentas_bancarias.id_persona = personas.id
GROUP BY
    facciones.nombre;

 * postgresql://postgres:***@localhost:5432/tarea1
6 rows affected.


[('JavaScriptia', 'Banco Scotiabank'),
 ('Pythonia', 'Banco Scotiabank'),
 ('MaCdonia', 'FedraBank S.A'),
 ('FEDRA', 'Banco Scotiabank'),
 ('Luciérnagas', 'FedraBank S.A'),
 ('Carroñeros', 'Banco Santander SpA')]

### Consulta 10 (6 pts)
**ID, nombre y segundo apellido de las personas traidoras**(Es decir, si la persona pertenece a dos o más facciones.)

Ejemplo:
   
|ID | Nombre | Apellido | 
|-|-------|-|
| 1 | Pedro | Pascal |
| 2 | Bella | Ramsey |
| ... | ... | ... | 

**Solución:** _Escriba aquí la descripción de su solución._

In [30]:
%%sql
SELECT
    personas.id "ID", personas.nombre "Nombre", personas.apellidos "Apellido"  
FROM personas
    INNER JOIN persona_pertenece_faccion ON personas.id = persona_pertenece_faccion.id_persona
GROUP BY personas.id
HAVING
    COUNT(personas.id) > 1;

 * postgresql://postgres:***@localhost:5432/tarea1
35 rows affected.


[(273, 'Frances', 'Russell Jackson'),
 (732, 'Diane', 'Parrett Rojas'),
 (554, 'Austin', 'Schaal Fernandez'),
 (946, 'Otto', 'Solt Smith'),
 (169, 'Michelle', 'Despard Christmas'),
 (605, 'Jill', 'Erb Gibbs'),
 (967, 'Nicole', 'Cauthen Johnson'),
 (870, 'Donna', 'Kelly Torres'),
 (957, 'Claudia', 'Haffey Warriner'),
 (764, 'Gino', 'Penrod Hopkins'),
 (940, 'Louis', 'Miller Dunn'),
 (815, 'John', 'Moczo Owens'),
 (681, 'Candice', 'Barbour Locks'),
 (974, 'Virginia', 'Bowling Gardner'),
 (631, 'Jeffrey', 'Thomas Geffrard'),
 (889, 'Daniel', 'Sexton Betts'),
 (692, 'Patricia', 'Fields Paul'),
 (637, 'Alejandro', 'Matin Cahill'),
 (7, 'Erma', 'Potter Daugherty'),
 (496, 'Linda', 'Remmers Miller'),
 (908, 'Claudia', 'Sheets Morin'),
 (71, 'Paul', 'Geiser Pelletier'),
 (514, 'Delores', 'Ellis Humble'),
 (798, 'Mary', 'Mcconnell Rosso'),
 (147, 'Delilah', 'Nardo Mirabal'),
 (338, 'Elsie', 'Ramirez Condon'),
 (419, 'Nancy', 'Humbert Smith'),
 (678, 'Francisco', 'Hendricks Mace'),
 (806, 'James', 'Morrison Cannon'),
 (287, 'Jody', 'Ulmer Hashim'),
 (586, 'Minerva', 'Reid Counts'),
 (442, 'David', 'King Delarosa'),
 (825, 'Elsa', 'Stumpf Dickson'),
 (332, 'Andrew', 'Scott Looney'),
 (215, 'Thomas', 'Plessinger Delgado')]

### Consulta BONUS (6 pts.)

**Nombre y Apellidos de las personas casadas**(solo se muestra una única vez la pareja)

Ejemplo:
   
|Nombre persona 1 | Apellidos persona 1 | Nombre persona 2 | Apellidos persona 2 |
|-|-------|-|-|
| Felipe | Martinez | Raul | Cuello |
| Rodrigo | Munita | Juan | Mira |
| Nicole | Rojel | Benjamin | Campos |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [72]:
%%sql
SELECT
    conyugue1.nombre, conyugue1.apellidos, conyugue2.nombre, conyugue2.apellidos
FROM personas conyugue1
JOIN personas conyugue2
ON conyugue1.conyugue = conyugue2.id
GROUP BY conyugue1.nombre, conyugue1.apellidos, conyugue2.nombre, conyugue2.apellidos;

 * postgresql://postgres:***@localhost:5432/tarea1
170 rows affected.


[('John', 'Schmelzer Griffin', 'Janet', 'Churchill Barber'),
 ('Jeffrey', 'Cocks Herring', 'Henry', 'Fisher Henning'),
 ('Jaime', 'Shelton Keitt', 'Tonia', 'Stuckey Martin'),
 ('Lucille', 'Cortez Williams', 'Heather', 'Hembree Mirabal'),
 ('Jason', 'Luster Escoto', 'Joyce', 'Learn Kimble'),
 ('Sherry', 'Long Brammell', 'Thomas', 'Smith Mumma'),
 ('Cory', 'Mabry Pinson', 'Fred', 'Ashmore Stoughton'),
 ('Rhea', 'Stiffler Butler', 'Bonnie', 'Espinal Baumgardner'),
 ('David', 'Horwath Mimes', 'John', 'Mcgrew Gutierrez'),
 ('Joyce', 'Learn Kimble', 'Jason', 'Luster Escoto'),
 ('Bridget', 'Wilkerson Mcfaddin', 'Robert', 'Kint Mccormick'),
 ('Sandra', 'Jack Hartfield', 'Frances', 'Russell Jackson'),
 ('Louis', 'Miller Dunn', 'Donald', 'Doherty Mason'),
 ('Jason', 'Foster Roley', 'John', 'Bailey Barron'),
 ('Francisco', 'Hendricks Mace', 'Tracy', 'Romero Stout'),
 ('Tammy', 'Laurie Washington', 'Michael', 'Estrada Gerke'),
 ('Rhonda', 'Alegre Green', 'Barbara', 'Leatherwood Wells'),
 ('David', 'Tep Wilson', 'Brian', 'Robinson Posthuma'),
 ('William', 'Knudtson Long', 'Marian', 'Clark Bartko'),
 ('Edward', 'Shaw Blaze', 'Deana', 'Long Craft'),
 ('Ernest', 'Buri Battaglia', 'Patricia', 'Campbell Dameron'),
 ('Ruth', 'Fletcher Bowman', 'Cynthia', 'Booker Scroggins'),
 ('Donald', 'Doherty Mason', 'Louis', 'Miller Dunn'),
 ('Marian', 'Clark Bartko', 'William', 'Knudtson Long'),
 ('Shirley', 'Koester Rabb', 'Mabel', 'Robinett Reis'),
 ('Ralph', 'Hanson Souder', 'Alice', 'Fife Roberts'),
 ('Ellen', 'Jenkins Gonzalez', 'Pamela', 'Davis Valdez'),
 ('Terrence', 'Coleman Nelson', 'Jose', 'Sliger Franco'),
 ('Shelby', 'Byrne Lee', 'Jeannie', 'Luke Szmidt'),
 ('Lee', 'Larkins Abbott', 'Tanya', 'Heiser Frazier'),
 ('Elliott', 'Hinkle Rodgers', 'Janice', 'Jones Trumbore'),
 ('Jill', 'Erb Gibbs', 'Eva', 'Ward Bailey'),
 ('Arthur', 'Sullivan Hu', 'Lydia', 'Encallado Peake'),
 ('Anna', 'Siebert Childs', 'Brandy', 'Williams Carter'),
 ('Sharon', 'Rainbolt Adams', 'Vance', 'Butler Williams'),
 ('Maria', 'Davis Webber', 'Rosa', 'Brown Williams'),
 ('Maria', 'Hodson Stubbert', 'Theodore', 'Canada Glover'),
 ('James', 'Morrison Cannon', 'Raymond', 'Coleman Jones'),
 ('Linda', 'Rimmer Westberry', 'Richard', 'Carroll Wainer'),
 ('Carly', 'Lassalle Swoope', 'Alexander', 'Mcclung Sanders'),
 ('John', 'Moon Fralin', 'Deidre', 'Ferguson Torres'),
 ('Sandra', 'Drumm Alsip', 'Marion', 'Govea Browning'),
 ('Helen', 'Morgan Chavez', 'Anthony', 'Nash Orourke'),
 ('Noah', 'Grunau Kaeo', 'Sandra', 'Ketcham Guglielmo'),
 ('Jason', 'Miller Hicks', 'George', 'Bourg Merrell'),
 ('James', 'Norman Smith', 'Jacinta', 'Timm Hanke'),
 ('William', 'Crawford Johnson', 'Michelle', 'Robinson Burchinal'),
 ('Cora', 'Ellison Long', 'Felicia', 'Moe Guidry'),
 ('Tammi', 'Stiles Rosenbloom', 'Deborah', 'Elmore Law'),
 ('Felicia', 'Moe Guidry', 'Cora', 'Ellison Long'),
 ('Frances', 'Russell Jackson', 'Sandra', 'Jack Hartfield'),
 ('David', 'John Martinez', 'Henry', 'Young Wing'),
 ('Pamela', 'Davis Valdez', 'Ellen', 'Jenkins Gonzalez'),
 ('Yolanda', 'Payne Davis', 'Heather', 'Brown Brewer'),
 ('Linda', 'Vinson Garcia', 'Paula', 'Sanchez Cooksey'),
 ('Nancy', 'Astorga Johnson', 'Peggy', 'Unterreiner Saulnier'),
 ('Harold', 'Schmiedeskamp Horton', 'Josphine', 'Grasso Sabol'),
 ('Tracy', 'Romero Stout', 'Francisco', 'Hendricks Mace'),
 ('Victoria', 'Williams Garnick', 'Roger', 'Zepeda Quintanilla'),
 ('George', 'Bourg Merrell', 'Jason', 'Miller Hicks'),
 ('Paula', 'Gillespie Mcclaughry', 'David', 'Dawson Miller'),
 ('John', 'Bailey Barron', 'Jason', 'Foster Roley'),
 ('Latasha', 'Luksa Pitchford', 'Nancy', 'Sevier Matthews'),
 ('Dinah', 'Carroll Medina', 'David', 'Crimmins White'),
 ('Henry', 'Young Wing', 'David', 'John Martinez'),
 ('Fred', 'Ashmore Stoughton', 'Cory', 'Mabry Pinson'),
 ('Angela', 'Coleman Franklin', 'Paul', 'Raether Brown'),
 ('David', 'Crimmins White', 'Dinah', 'Carroll Medina'),
 ('Edward', 'Backes Talley', 'Vernon', 'George Mieloszyk'),
 ('Paul', 'Hyre Falgout', 'Arnul